src/apps/common/submodules/portWizard/portWizard.js

```diff
@@ -505,6 +593,68 @@ define(function(require) {
 			};
 		},
 
+		/**
+		 * Get the losing and winning carriers for the phone numbers
+		 * @param  {Object} args
+		 * @param  {String} args.numbers  Phone numbers to look up
+		 * @param  {Function} args.success  Success callback
+		 * @param  {Function} [args.error]  Error callback
+		 */
+		portWizardCarrierSelectionGetNumberCarriers: function(args) {
+			var self = this,
+				numbers = args.numbers;
+
+			monster.waterfall([
+				function(waterfallCallback) {
+					self.portWizardRequestPhoneNumbersLookup({
+						numbers: numbers,
+						success: function(data) {
+							waterfallCallback(null, data.numbers);
+						},
+						error: function() {
+							waterfallCallback(true);
+						}
+					});
+				},
+				function(numbersData, waterfallCallback) {
+					var numbersByLosingCarrier = _
+							.groupBy(numbers, function(number) {
+								return _.get(numbersData, [
+									number,
+									'losing_carrier',
+									'name'
+								], 'Unknown');
+							}),
+						winningCarriers = _
+							.chain(numbersData)
+							.map('carriers')
+							.reduce(function(commonCarriers, numberCarriers) {
+								var portableCarriers = _
+									.chain(numberCarriers)
+									.filter('portability')
+									.map('name')
+									.value();
+
+								return _.isEmpty(commonCarriers)
+									? portableCarriers
+									: _.intersection(commonCarriers, portableCarriers);
+							}, [])
+							.value();
```

Comment:
This piece of logic seems a bit convoluted as well; as I understand it, it returns an array containing carriers allowing porting that are common to all numbers.

To make this more clear, we could extract the filtering/mapping from the reduce:
```js
_.map(numbersData, function(numberData) {
    return _
        .chain(numberData.carriers)
        .filter('portability')
        .map('name')
        .value()
})
```
This would return an array of arrays, each containing the list of carriers allowing porting for a specific number. We could then spread that into `lodash#intersection` to retrieve the common carriers:
```js
_
.chain(numbersData)
.map(function(numberData) {
    return _
        .chain(numberData.carriers)
        .filter('portability')
        .map('name')
        .value()
})
.thru(_.spread(_.intersection))
.value()
```
Now this is somewhat easier to follow but still kind of hard to understand without context, so we could use a named function expression for `_.spread(_.intersection)` and give it a more meaningful name:
```js
var findCommonCarriers = _.spread(_.intersection);
_
.chain(numbersData)
.map(function(numberData) {
    return _
        .chain(numberData.carriers)
        .filter('portability')
        .map('name')
        .value()
})
.thru(findCommonCarriers)
.value()
```
As always, let me know if you have any questions for me regarding this!

Reply:
That is nice! It will be a lot more readable that way.
Thanks also for the suggestion on the `lodash#spread` usage, I was wondering how to pass the array items as parameters for the `lodash#intersection` function and similar ones.